File: backend/src/advisory/crop-recommendation.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { CropRecommendationRequestDto, CropRecommendationDto } from './dto/advisory.dto';

// Crop database with suitability parameters
const CROP_DATABASE = [
  { name: 'Rice', variety: 'Basmati', optimalPh: [5.5, 7.0], waterReq: 'high', season: 'kharif', baseYield: 25, baseCost: 35000, baseRevenue: 75000 },
  { name: 'Rice', variety: 'Sona Masuri', optimalPh: [5.5, 7.0], waterReq: 'high', season: 'kharif', baseYield: 28, baseCost: 32000, baseRevenue: 70000 },
  { name: 'Wheat', variety: 'HD-2967', optimalPh: [6.0, 7.5], waterReq: 'medium', season: 'rabi', baseYield: 20, baseCost: 25000, baseRevenue: 50000 },
  { name: 'Maize', variety: 'Hybrid', optimalPh: [5.8, 7.0], waterReq: 'medium', season: 'kharif', baseYield: 30, baseCost: 28000, baseRevenue: 55000 },
  { name: 'Cotton', variety: 'Bt Cotton', optimalPh: [6.0, 8.0], waterReq: 'medium', season: 'kharif', baseYield: 8, baseCost: 40000, baseRevenue: 80000 },
  { name: 'Sugarcane', variety: 'Co-0238', optimalPh: [6.0, 7.5], waterReq: 'high', season: 'annual', baseYield: 350, baseCost: 60000, baseRevenue: 140000 },
  { name: 'Groundnut', variety: 'TMV-2', optimalPh: [6.0, 7.0], waterReq: 'low', season: 'kharif', baseYield: 12, baseCost: 30000, baseRevenue: 60000 },
  { name: 'Soybean', variety: 'JS-335', optimalPh: [6.0, 7.0], waterReq: 'medium', season: 'kharif', baseYield: 10, baseCost: 22000, baseRevenue: 45000 },
  { name: 'Tomato', variety: 'Hybrid', optimalPh: [6.0, 7.0], waterReq: 'medium', season: 'rabi', baseYield: 150, baseCost: 45000, baseRevenue: 120000 },
  { name: 'Onion', variety: 'Nasik Red', optimalPh: [6.0, 7.0], waterReq: 'medium', season: 'rabi', baseYield: 100, baseCost: 35000, baseRevenue: 80000 },
];

@Injectable()
export class CropRecommendationService {
  private readonly logger = new Logger(CropRecommendationService.name);

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService,
  ) {}

  async generateRecommendations(
    userId: string,
    dto: CropRecommendationRequestDto,
  ): Promise<CropRecommendationDto[]> {
    const { location, soilData, preferences, parcelId } = dto;

    // Get historical data if parcel is provided
    let cropHistory: any[] = [];
    if (parcelId) {
      cropHistory = await this.prisma.cropHistory.findMany({
        where: { parcelId },
        orderBy: { sowingDate: 'desc' },
        take: 5,
      });
    }

    // Calculate suitability scores for each crop
    const recommendations = CROP_DATABASE.map((crop) => {
      let score = 50; // Base score
      const reasoning: string[] = [];
      const risks: string[] = [];

      // pH suitability (0-30 points)
      const phScore = this.calculatePhScore(soilData.ph, crop.optimalPh as [number, number]);
      score += phScore;
      if (phScore > 20) {
        reasoning.push(`Soil pH (${soilData.ph}) is optimal for ${crop.name}`);
      } else if (phScore < 10) {
        risks.push(`Soil pH (${soilData.ph}) is not ideal for ${crop.name}`);
      }

      // Nutrient availability (0-20 points)
      const nutrientScore = this.calculateNutrientScore(soilData);
      score += nutrientScore;
      if (nutrientScore > 15) {
        reasoning.push('Good nutrient levels in soil');
      } else if (nutrientScore < 8) {
        risks.push('Nutrient deficiency may affect yield');
      }

      // Preference bonus (0-10 points)
      if (preferences?.includes(crop.name)) {
        score += 10;
        reasoning.push('Matches your crop preference');
      }

      // Historical performance adjustment
      const historicalCrop = cropHistory.find((h) => h.cropName === crop.name);
      if (historicalCrop) {
        if (historicalCrop.yield && historicalCrop.yield > crop.baseYield * 0.8) {
          score += 5;
          reasoning.push('Good historical performance on this land');
        } else if (historicalCrop.yield && historicalCrop.yield < crop.baseYield * 0.5) {
          score -= 10;
          risks.push('Poor historical performance on this land');
        }
      }

      // Normalize score to 0-100
      score = Math.max(0, Math.min(100, score));

      // Adjust yield and revenue based on soil conditions
      const yieldMultiplier = 0.7 + (score / 100) * 0.6;
      const expectedYield = crop.baseYield * yieldMultiplier;
      const estimatedRevenue = crop.baseRevenue * yieldMultiplier;

      return {
        cropName: crop.name,
        variety: crop.variety,
        suitabilityScore: Math.round(score),
        expectedYield: Math.round(expectedYield * 10) / 10,
        estimatedInputCost: crop.baseCost,
        estimatedRevenue: Math.round(estimatedRevenue),
        reasoning,
        risks,
      };
    });

    // Sort by suitability score (descending)
    recommendations.sort((a, b) => b.suitabilityScore - a.suitabilityScore);

    // Return top 5 recommendations
    const topRecommendations = recommendations.slice(0, 5);

    // Store recommendation in database
    await this.storeRecommendation(userId, parcelId, topRecommendations);

    this.logger.log(`Generated ${topRecommendations.length} recommendations for user ${userId}`);

    return topRecommendations;
  }

  private calculatePhScore(ph: number, optimalRange: [number, number]): number {
    const [min, max] = optimalRange;
    if (ph >= min && ph <= max) {
      return 30; // Perfect pH
    }
    const distance = ph < min ? min - ph : ph - max;
    return Math.max(0, 30 - distance * 10);
  }

  private calculateNutrientScore(soilData: { nitrogen: number; phosphorus: number; potassium: number }): number {
    let score = 0;
    
    // Nitrogen (optimal: 250-350 kg/ha)
    if (soilData.nitrogen >= 250 && soilData.nitrogen <= 350) score += 7;
    else if (soilData.nitrogen >= 150) score += 4;
    
    // Phosphorus (optimal: 20-40 kg/ha)
    if (soilData.phosphorus >= 20 && soilData.phosphorus <= 40) score += 7;
    else if (soilData.phosphorus >= 10) score += 4;
    
    // Potassium (optimal: 150-250 kg/ha)
    if (soilData.potassium >= 150 && soilData.potassium <= 250) score += 6;
    else if (soilData.potassium >= 100) score += 3;

    return score;
  }

  private async storeRecommendation(userId: string, parcelId: string | undefined, recommendations: CropRecommendationDto[]) {
    if (!parcelId) return;

    try {
      await this.prisma.cropRecommendation.create({
        data: {
          user: { connect: { id: userId } },
          parcel: { connect: { id: parcelId } },
          recommendedCrops: {
            create: recommendations.map((r) => ({
              cropName: r.cropName,
              variety: r.variety,
              suitabilityScore: r.suitabilityScore,
              expectedYield: r.expectedYield,
              estimatedInputCost: r.estimatedInputCost,
              estimatedRevenue: r.estimatedRevenue,
              reasoning: r.reasoning,
              risks: r.risks,
            })),
          },
        },
      });
    } catch (error) {
      this.logger.error('Failed to store recommendation', error);
    }
  }
}

File: ml-service/services/crop_recommendation_service.py
"""Crop recommendation service with ensemble model inference."""
import logging
from typing import List, Optional, Dict, Any
import numpy as np

from schemas import (
    CropRecommendationRequest, CropRecommendationResponse,
    CropRecommendation, SoilData, WeatherContext
)
from model_loader import get_model_loader
from config import settings

logger = logging.getLogger(__name__)

# Crop database with suitability parameters (used for fallback and enrichment)
CROP_DATABASE = {
    "rice": {
        "soil_types": ["clay", "loamy", "alluvial"],
        "ph_range": (5.5, 7.5),
        "water_requirement": "high",
        "season": ["kharif", "monsoon"],
        "npk_preference": {"n": "high", "p": "medium", "k": "medium"},
        "base_yield": 4500,
        "risks": ["drought", "flooding", "blast_disease"]
    },
    "wheat": {
        "soil_types": ["loamy", "clay", "black"],
        "ph_range": (6.0, 7.5),
        "water_requirement": "medium",
        "season": ["rabi", "winter"],
        "npk_preference": {"n": "high", "p": "medium", "k": "low"},
        "base_yield": 3500,
        "risks": ["rust", "heat_stress", "lodging"]
    },
    "maize": {
        "soil_types": ["loamy", "sandy", "alluvial"],
        "ph_range": (5.5, 7.5),
        "water_requirement": "medium",
        "season": ["kharif", "rabi"],
        "npk_preference": {"n": "high", "p": "medium", "k": "medium"},
        "base_yield": 5000,
        "risks": ["stem_borer", "drought", "fall_armyworm"]
    },
    "cotton": {
        "soil_types": ["black", "loamy", "alluvial"],
        "ph_range": (6.0, 8.0),
        "water_requirement": "medium",
        "season": ["kharif"],
        "npk_preference": {"n": "medium", "p": "high", "k": "high"},
        "base_yield": 500,
        "risks": ["bollworm", "whitefly", "water_stress"]
    },
    "sugarcane": {
        "soil_types": ["loamy", "clay", "alluvial"],
        "ph_range": (6.0, 8.0),
        "water_requirement": "high",
        "season": ["annual"],
        "npk_preference": {"n": "high", "p": "medium", "k": "high"},
        "base_yield": 70000,
        "risks": ["red_rot", "water_logging", "termites"]
    },
    "groundnut": {
        "soil_types": ["sandy", "loamy", "red"],
        "ph_range": (6.0, 7.0),
        "water_requirement": "low",
        "season": ["kharif", "rabi"],
        "npk_preference": {"n": "low", "p": "high", "k": "medium"},
        "base_yield": 1500,
        "risks": ["leaf_spot", "collar_rot", "drought"]
    },
    "soybean": {
        "soil_types": ["loamy", "clay", "black"],
        "ph_range": (6.0, 7.5),
        "water_requirement": "medium",
        "season": ["kharif"],
        "npk_preference": {"n": "low", "p": "high", "k": "medium"},
        "base_yield": 2000,
        "risks": ["rust", "pod_borer", "yellow_mosaic"]
    },
    "tomato": {
        "soil_types": ["loamy", "sandy", "red"],
        "ph_range": (6.0, 7.0),
        "water_requirement": "medium",
        "season": ["rabi", "summer"],
        "npk_preference": {"n": "medium", "p": "high", "k": "high"},
        "base_yield": 25000,
        "risks": ["blight", "fruit_borer", "wilting"]
    },
    "potato": {
        "soil_types": ["loamy", "sandy", "alluvial"],
        "ph_range": (5.5, 6.5),
        "water_requirement": "medium",
        "season": ["rabi"],
        "npk_preference": {"n": "high", "p": "high", "k": "high"},
        "base_yield": 20000,
        "risks": ["late_blight", "early_blight", "tuber_moth"]
    },
    "onion": {
        "soil_types": ["loamy", "sandy", "alluvial"],
        "ph_range": (6.0, 7.0),
        "water_requirement": "medium",
        "season": ["rabi", "kharif"],
        "npk_preference": {"n": "medium", "p": "medium", "k": "medium"},
        "base_yield": 15000,
        "risks": ["purple_blotch", "thrips", "storage_rot"]
    },
    "chickpea": {
        "soil_types": ["loamy", "clay", "black"],
        "ph_range": (6.0, 8.0),
        "water_requirement": "low",
        "season": ["rabi"],
        "npk_preference": {"n": "low", "p": "high", "k": "medium"},
        "base_yield": 1200,
        "risks": ["wilt", "pod_borer", "ascochyta_blight"]
    },
    "lentil": {
        "soil_types": ["loamy", "clay", "alluvial"],
        "ph_range": (6.0, 8.0),
        "water_requirement": "low",
        "season": ["rabi"],
        "npk_preference": {"n": "low", "p": "high", "k": "medium"},
        "base_yield": 1000,
        "risks": ["rust", "wilt", "aphids"]
    },
    "mustard": {
        "soil_types": ["loamy", "sandy", "alluvial"],
        "ph_range": (6.0, 7.5),
        "water_requirement": "low",
        "season": ["rabi"],
        "npk_preference": {"n": "medium", "p": "medium", "k": "low"},
        "base_yield": 1500,
        "risks": ["aphids", "white_rust", "alternaria_blight"]
    },
    "sunflower": {
        "soil_types": ["loamy", "clay", "black"],
        "ph_range": (6.0, 7.5),
        "water_requirement": "medium",
        "season": ["kharif", "rabi"],
        "npk_preference": {"n": "medium", "p": "high", "k": "medium"},
        "base_yield": 1800,
        "risks": ["head_rot", "rust", "alternaria"]
    },
    "millet": {
        "soil_types": ["sandy", "loamy", "red"],
        "ph_range": (5.5, 7.5),
        "water_requirement": "low",
        "season": ["kharif"],
        "npk_preference": {"n": "low", "p": "medium", "k": "low"},
        "base_yield": 1500,
        "risks": ["downy_mildew", "blast", "stem_borer"]
    }
}


class CropRecommendationService:
    """Service for generating crop recommendations."""
    
    def __init__(self):
        self.model_loader = get_model_loader()
        self._feature_cols = ['N', 'P', 'K', 'temperature', 'humidity', 'ph', 'rainfall']
    
    def _extract_features(self, request: CropRecommendationRequest) -> np.ndarray:
        """Extract features from request for model inference, mapping to physical constraints."""
        soil = request.soil_data
        weather = request.weather_context or WeatherContext()
        
        # Enforce realistic constraints to prevent model hallucination on extreme outliers
        n = max(0.0, min(float(soil.nitrogen), 300.0))
        p = max(0.0, min(float(soil.phosphorus), 150.0))
        k = max(0.0, min(float(soil.potassium), 300.0))
        temp = max(-10.0, min(float(weather.temperature or 25.0), 60.0))
        hum = max(0.0, min(float(weather.humidity or 60.0), 100.0))
        ph = max(0.0, min(float(soil.ph), 14.0))
        rain = max(0.0, min(float(weather.rainfall or 100.0), 1000.0))
        
        features = [n, p, k, temp, hum, ph, rain]
        
        return np.array([features], dtype=np.float32)
    
    def _calculate_soil_suitability(self, crop_info: Dict, soil: SoilData) -> float:
        """Calculate soil suitability score for a crop."""
        score = 0.0
        
        # Soil type match (40 points)
        if soil.type.lower() in crop_info["soil_types"]:
            score += 40.0
        else:
            score += 10.0
        
        # pH suitability (30 points)
        ph_min, ph_max = crop_info["ph_range"]
        if ph_min <= soil.ph <= ph_max:
            score += 30.0
        elif abs(soil.ph - ph_min) <= 0.5 or abs(soil.ph - ph_max) <= 0.5:
            score += 15.0
        else:
            score += 5.0
        
        # NPK suitability (30 points)
        npk_score = 0.0
        npk_pref = crop_info["npk_preference"]
        
        if npk_pref["n"] == "high" and soil.nitrogen >= 200:
            npk_score += 10.0
        elif npk_pref["n"] == "medium" and 100 <= soil.nitrogen < 200:
            npk_score += 10.0
        elif npk_pref["n"] == "low" and soil.nitrogen < 100:
            npk_score += 10.0
        else:
            npk_score += 5.0
        
        if npk_pref["p"] == "high" and soil.phosphorus >= 50:
            npk_score += 10.0
        elif npk_pref["p"] == "medium" and 25 <= soil.phosphorus < 50:
            npk_score += 10.0
        else:
            npk_score += 5.0
        
        if npk_pref["k"] == "high" and soil.potassium >= 200:
            npk_score += 10.0
        elif npk_pref["k"] == "medium" and 100 <= soil.potassium < 200:
            npk_score += 10.0
        elif npk_pref["k"] == "low" and soil.potassium < 100:
            npk_score += 10.0
        else:
            npk_score += 5.0
        
        score += npk_score
        return score
    
    def _generate_reasoning(self, crop_name: str, crop_info: Dict, soil: SoilData, 
                           model_confidence: Optional[float] = None) -> List[str]:
        """Generate reasoning for crop recommendation."""
        reasoning = []
        
        if model_confidence:
            reasoning.append(f"ML model confidence: {model_confidence:.1%}")
        
        if soil.type.lower() in crop_info["soil_types"]:
            reasoning.append(f"{soil.type.capitalize()} soil is well-suited for {crop_name}")
        
        ph_min, ph_max = crop_info["ph_range"]
        if ph_min <= soil.ph <= ph_max:
            reasoning.append(f"Soil pH {soil.ph} is optimal (ideal: {ph_min}-{ph_max})")
        
        if soil.nitrogen >= 150:
            reasoning.append("Good nitrogen levels support healthy growth")
        
        if soil.phosphorus >= 40:
            reasoning.append("Adequate phosphorus for root development")
        
        return reasoning if reasoning else [f"{crop_name} can be grown in current conditions"]
    
    def _get_risks(self, crop_info: Dict, weather: Optional[WeatherContext]) -> List[str]:
        """Get relevant risks for the crop."""
        risks = []
        base_risks = crop_info.get("risks", [])
        
        if weather:
            if weather.rainfall and weather.rainfall < 50:
                if "drought" in base_risks:
                    risks.append("Low rainfall may cause drought stress")
            if weather.temperature and weather.temperature > 35:
                if "heat_stress" in base_risks:
                    risks.append("High temperatures may cause heat stress")
        
        for risk in base_risks[:2]:
            risk_text = risk.replace("_", " ").title()
            if risk_text not in [r.split()[0] for r in risks]:
                risks.append(f"Monitor for {risk_text}")
        
        return risks
    
    def _run_model_inference(self, features: np.ndarray) -> Optional[Dict[str, float]]:
        """Run inference using the loaded ensemble model."""
        model_data = self.model_loader.get_model("crop_recommendation")
        
        if model_data is not None:
            try:
                # Handle dictionary format from training
                if isinstance(model_data, dict):
                    model = model_data['model']
                    scaler = model_data['scaler']
                    crops = model_data.get('classes', [])
                    
                    # Scale features
                    features_scaled = scaler.transform(features)
                    
                    # Get probabilities
                    proba = model.predict_proba(features_scaled)
                    
                    if crops and len(proba[0]) == len(crops):
                        result = {crop: float(proba[0][i]) for i, crop in enumerate(crops)}
                        logger.info(f"Model predictions: {sorted(result.items(), key=lambda x: x[1], reverse=True)[:5]}")
                        return result
                
                # Handle object with predict_proba method
                elif hasattr(model_data, 'predict_proba'):
                    proba = model_data.predict_proba(features)
                    crops = model_data.get_all_crops() if hasattr(model_data, 'get_all_crops') else self.model_loader.get_crop_labels()
                    
                    if crops and len(proba[0]) == len(crops):
                        result = {crop: float(proba[0][i]) for i, crop in enumerate(crops)}
                        logger.info(f"Model predictions: {sorted(result.items(), key=lambda x: x[1], reverse=True)[:5]}")
                        return result
                
            except Exception as e:
                logger.error(f"Model inference failed: {e}")
                import traceback
                logger.error(traceback.format_exc())
                return None
        return None
    
    def _run_fallback_recommendation(self, request: CropRecommendationRequest) -> List[CropRecommendation]:
        """Generate recommendations using rule-based approach."""
        recommendations = []
        
        for crop_name, crop_info in CROP_DATABASE.items():
            soil_score = self._calculate_soil_suitability(crop_info, request.soil_data)
            
            weather_adjustment = 0.0
            if request.weather_context:
                weather = request.weather_context
                water_req = crop_info["water_requirement"]
                
                if water_req == "high" and weather.rainfall and weather.rainfall >= 150:
                    weather_adjustment += 10.0
                elif water_req == "low" and weather.rainfall and weather.rainfall < 100:
                    weather_adjustment += 10.0
                elif water_req == "medium":
                    weather_adjustment += 5.0
            
            final_score = min(100.0, soil_score + weather_adjustment)
            base_yield = crop_info["base_yield"]
            yield_factor = final_score / 100.0
            expected_yield = base_yield * yield_factor * 0.9
            
            recommendation = CropRecommendation(
                crop_name=crop_name,
                variety="Standard",
                suitability_score=round(final_score, 2),
                expected_yield=round(expected_yield, 2),
                reasoning=self._generate_reasoning(crop_name, crop_info, request.soil_data),
                risks=self._get_risks(crop_info, request.weather_context)
            )
            recommendations.append(recommendation)
        
        recommendations.sort(key=lambda x: x.suitability_score, reverse=True)
        return recommendations[:5]
    
    def get_recommendations(self, request: CropRecommendationRequest) -> CropRecommendationResponse:
        """Generate crop recommendations based on input data."""
        model_info = self.model_loader.get_model_info("crop_recommendation")
        model_version = model_info.version if model_info else "fallback"
        is_fallback = False
        fallback_disclaimer = None
        
        # Try model inference first
        features = self._extract_features(request)
        model_predictions = self._run_model_inference(features)
        
        if model_predictions:
            recommendations = []
            for crop_name, prob in model_predictions.items():
                crop_info = CROP_DATABASE.get(crop_name, {
                    "soil_types": ["loamy"],
                    "ph_range": (6.0, 7.5),
                    "water_requirement": "medium",
                    "npk_preference": {"n": "medium", "p": "medium", "k": "medium"},
                    "base_yield": 2000,
                    "risks": []
                })
                
                soil_score = self._calculate_soil_suitability(crop_info, request.soil_data)
                # Combine model prediction with rule-based score
                final_score = min(100.0, (prob * 60) + (soil_score * 0.4))
                
                recommendation = CropRecommendation(
                    crop_name=crop_name,
                    variety="Standard",
                    suitability_score=round(final_score, 2),
                    expected_yield=round(crop_info["base_yield"] * (final_score / 100), 2),
                    reasoning=self._generate_reasoning(crop_name, crop_info, request.soil_data, prob),
                    risks=self._get_risks(crop_info, request.weather_context)
                )
                recommendations.append(recommendation)
            
            recommendations.sort(key=lambda x: x.suitability_score, reverse=True)
            recommendations = recommendations[:5]
            
            if model_info and model_info.metadata.get("trained_at"):
                model_version = f"trained-{model_info.version}"
        else:
            is_fallback = True
            fallback_disclaimer = "Using rule-based recommendations. Results may vary."
            model_version = "rule-based-1.0.0"
            recommendations = self._run_fallback_recommendation(request)
        
        return CropRecommendationResponse(
            recommendations=recommendations,
            model_version=model_version,
            is_fallback=is_fallback,
            fallback_disclaimer=fallback_disclaimer
        )


# Singleton instance
_crop_recommendation_service: Optional[CropRecommendationService] = None


def get_crop_recommendation_service() -> CropRecommendationService:
    """Get the crop recommendation service instance."""
    global _crop_recommendation_service
    if _crop_recommendation_service is None:
        _crop_recommendation_service = CropRecommendationService()
    return _crop_recommendation_service
